scio-smb/src/main/java/org/apache/beam/sdk/extensions/smb/SortedBucketTransform.java

```diff
@@ -72,56 +72,72 @@
  * @param <FinalValueT>
  */
 public class SortedBucketTransform<FinalKeyT, FinalValueT> extends PTransform<PBegin, WriteResult> {
+  private static final Logger LOG = LoggerFactory.getLogger(SortedBucketTransform.class);
   // Dataflow calls split() with a suggested byte size that assumes a higher throughput than
   // SMB joins have. By adjusting this suggestion we can arrive at a more optimal parallelism.
   static final Double DESIRED_SIZE_BYTES_ADJUSTMENT_FACTOR = 0.33;
 
-  private final MergeAndWriteBucketsSource<FinalKeyT, FinalValueT> boundedSource;
-  private final FinalizeTransformedBuckets<FinalValueT> finalizeBuckets;
+  private final BucketSource<FinalKeyT> bucketSource;
+  private final DoFn<Iterable<MergedBucket>, KV<BucketShardId, ResourceId>> finalizeBuckets;
+  private final ParDo.SingleOutput<BucketItem, MergedBucket> doFn;
 
   public SortedBucketTransform(
       Class<FinalKeyT> finalKeyClass,
-      List<BucketedInput<?, ?>> sources,
+      List<SortedBucketSource.BucketedInput<?, ?>> sources,
       TargetParallelism targetParallelism,
       TransformFn<FinalKeyT, FinalValueT> transformFn,
+      TransformFnWithSideInputContext<FinalKeyT, FinalValueT> sideInputTransformFn,
       ResourceId outputDirectory,
       ResourceId tempDirectory,
+      Iterable<PCollectionView<?>> sides,
       NewBucketMetadataFn<FinalKeyT, FinalValueT> newBucketMetadataFn,
       FileOperations<FinalValueT> fileOperations,
       String filenameSuffix,
       String filenamePrefix) {
+    assert !((transformFn == null) && (sideInputTransformFn == null)); // at least one defined
+    assert !((transformFn != null) && (sideInputTransformFn != null)); // only one defined
+    assert sides == null || (sideInputTransformFn != null);
+
     final SMBFilenamePolicy filenamePolicy =
         new SMBFilenamePolicy(outputDirectory, filenamePrefix, filenameSuffix);
     final SourceSpec<FinalKeyT> sourceSpec = SourceSpec.from(finalKeyClass, sources);
-
-    boundedSource =
-        new MergeAndWriteBucketsSource<>(
-            finalKeyClass,
-            sources,
-            targetParallelism,
-            1,
-            0,
-            sourceSpec,
-            Metrics.distribution(getName(), getName() + "-KeyGroupSize"),
-            -1,
-            filenamePolicy.forTempFiles(tempDirectory),
-            fileOperations,
-            transformFn);
-
+    bucketSource = new BucketSource<>(sources, targetParallelism, 1, 0, sourceSpec, -1);
     finalizeBuckets =
         new FinalizeTransformedBuckets<>(
             fileOperations,
             newBucketMetadataFn,
             filenamePolicy.forDestination(),
             sourceSpec.hashType);
+
+    final FileAssignment fileAssignment = filenamePolicy.forTempFiles(tempDirectory);
+    final Distribution dist = Metrics.distribution(getName(), getName() + "-KeyGroupSize");
+    if (transformFn != null) {
+      this.doFn =
+          ParDo.of(
+              new NoSides<>(
```

Comment:
for the sake of a clearer job graph, maybe rename to `TransformWithNoSides`/`TransformWithSides` ?

Reply:
Sounds reasonable!